### PR TITLE
 Fixes #248 solubility as table 

### LINKED ISSUE #248: Define solubility as a table (X=pH, Y = Sol)
* Values can be edited by hand
* Values can be loaded from file

similar to Table Formulation UI

### PR DESCRIPTION
**Pull request by msevestre**

### CODE REVIEW

**Comment by msevestre** [comment]
@Yuri05 Big pull request that should basically support Table solubility in PK-Sim
There is a test that creates a simulation using Table Solubility. This test WILL fail as long as the implementation is not supported in SimModel.

**Comment by msevestre** [comment]
@Yuri05 
Table is exported as follow:
```
 "UseDerivedValues": true,
        "Points": [
          {
            "X": 5.0,
            "Y": 22.0,
            "RestartSolver": false
          },
          {
            "X": 10.0,
            "Y": 50.0,
            "RestartSolver": false
          }

```

Is this correct (I do not know what `UseDerivedValues` means anymore)

**Comment by Yuri05** [comment]
>Is this correct

No, must be `"UseDerivedValues": false`
Otherwise dY/dX (here: dSolubility/dpH) will be used in the simulation

File: src/PKSim.Assets/PKSimConstants.cs [hunk 1]
```diff
@@ -2157,6 +2160,7 @@ public static class UI
          public static readonly string UseWatermark = "Use watermark in charts when exporting to clipboard?";
          public static readonly string WatermarkText = "Text";
          public static readonly string WatermarkProperties = "Watermark Properties";
+         public static readonly string CreateTableSolubilityAlternative = "Create as pH-Sol table";
```
Comment (Yuri05) [inline src/PKSim.Assets/PKSimConstants.cs]: Think _Create as pH-Sol**ubility** table_ would be better

File: src/PKSim.Core/Model/FormulaFactory.cs [hunk 2]
```diff
@@ -311,6 +311,9 @@ private IFormula createFormula(RateKey rateKey)
          if (rateKey.IsTableWithOffsetFormula)
```
[the remark this answers is not in the document]
Comment (Yuri05) [reply src/PKSim.Core/Model/FormulaFactory.cs]: might be required for old projects (unless we don't convert old **TableWithOffset** formulas into **TableWithXArg**)

File: src/PKSim.Assets/PKSimConstants.cs (hunk 1 above)
Comment (msevestre) [reply src/PKSim.Assets/PKSimConstants.cs]: :)
image

changed

File: src/PKSim.Core/Model/FormulaFactory.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (msevestre) [reply src/PKSim.Core/Model/FormulaFactory.cs]: No this is only used when creating  a new Formula using the RateKey (e..g from DB).
That should never happen anymore. TableWithOffset will still be available as an object for backward compatibility

File: src/PKSim.Core/Model/FormulaFactory.cs [hunk 3]
```diff
@@ -348,6 +351,27 @@ private IFormula createFormula(RateKey rateKey)
...
+         formula.AddTableObjectPath(tableObjectPath);
+         formula.AddXArgumentObjectPath(xArgument);
+
+         //Table formula with reference has the same dimension as its referenced table object
+         formula.Dimension = tableObjectPath.Dimension;
+
+         return formula;
+      }
+
       private IFormula createTableFormulaWithOffset(RateKey rateKey)
```
Comment (msevestre) [inline src/PKSim.Core/Model/FormulaFactory.cs]: All of that should go to

File: src/PKSim.Assets/PKSimConstants.cs (hunk 1 above)
Comment (msevestre) [reply src/PKSim.Assets/PKSimConstants.cs]: done

File: src/PKSim.Assets/PKSimConstants.cs [hunk 4]
```diff
@@ -2238,13 +2238,25 @@ public static string ImportFormulationDescription
...
+               sb.AppendLine("The Excel table for formulation import should have <b>two colums</b>.");
+               sb.AppendLine(" -  The first column represents the Time (unit Time e.g. in hours).");
+               sb.AppendLine(" -  The second column represents the fraction of the dose released at t.");
                return sb.ToString();               
             }
-         } 
-         
+         }
+
+         public static string ImportSolubilityTableDescription
```
Comment (msevestre) [inline src/PKSim.Assets/PKSimConstants.cs]: Added a description too

File: src/PKSim.Assets/PKSimConstants.cs [hunk 5]
```diff
@@ -2238,13 +2238,25 @@ public static string ImportFormulationDescription
             get
             {
                var sb = new StringBuilder();
-               sb.AppendLine("The Excel table for formulation import should have a format of two colums.");
```
Comment (msevestre) [inline src/PKSim.Assets/PKSimConstants.cs]: this was not not english

File: src/PKSim.Core/Model/FormulaFactory.cs [hunk 6]
```diff
@@ -260,9 +261,10 @@ public IFormula RateFor(RateKey rateKey, IFormulaCache formulaCache)
          return formulaCache[rateKey];
       }
 
-      public TableFormula CreateTableFormula()
+      public TableFormula CreateTableFormula(bool useDerivedValues = true)
```
Comment (msevestre) [inline src/PKSim.Core/Model/FormulaFactory.cs]: add the parameter. Default is true (same behavior as previous)

File: src/PKSim.Core/Model/ParameterAlternativeFactory.cs [hunk 7]
```diff
@@ -59,11 +59,10 @@ public ParameterAlternative CreateTableAlternativeFor(ParameterAlternativeGroup
          var alternative = CreateAlternativeFor(compoundParameterGroup);
 
          var tableParameter = alternative.Parameter(tableParameterName);
-         var tableFormula = _formulaFactory.CreateTableFormula();
+         var tableFormula = _formulaFactory.CreateTableFormula(useDerivedValues:false);
```
Comment (msevestre) [inline src/PKSim.Core/Model/ParameterAlternativeFactory.cs]: creates with false in this case

**Pull request merged**